File: src/jade/mcp/memory_server.py
"""FastMCP knowledge graph memory server.

Implements the 9-tool MCP memory protocol with JSONL file persistence.
Based on @modelcontextprotocol/server-memory patterns.
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass, field
from typing import Any

from mcp.server.fastmcp import FastMCP


@dataclass
class KnowledgeGraph:
    """In-memory knowledge graph with JSONL persistence."""

    entities: list[dict[str, Any]] = field(default_factory=list)
    relations: list[dict[str, Any]] = field(default_factory=list)
    file_path: str = "./memory.jsonl"

    def load(self) -> None:
        """Load graph from JSONL file."""
        if not os.path.exists(self.file_path):
            return
        try:
            with open(self.file_path, encoding="utf-8") as f:
                for line in f:
                    line = line.strip()
                    if not line:
                        continue
                    try:
                        record = json.loads(line)
                    except json.JSONDecodeError:
                        continue  # Skip corrupt lines, keep loading valid ones
                    if record.get("type") == "entity":
                        # Avoid duplicates
                        existing = [e for e in self.entities if e["name"] == record["name"]]
                        if not existing:
                            self.entities.append(record)
                        else:
                            # Merge observations (preserve insertion order)
                            merged = existing[0].get("observations", []) + record.get("observations", [])
                            existing[0]["observations"] = list(dict.fromkeys(merged))
                    elif record.get("type") == "relation":
                        self.relations.append(record)
        except OSError:
            pass  # File unreadable — start with empty graph

    def save(self) -> None:
        """Save graph to JSONL file."""
        with open(self.file_path, "w", encoding="utf-8") as f:
            for entity in self.entities:
                record = {**entity, "type": "entity"}
                f.write(json.dumps(record) + "\n")
            for relation in self.relations:
                record = {**relation, "type": "relation"}
                f.write(json.dumps(record) + "\n")

    def find_entity(self, name: str) -> dict[str, Any] | None:
        for e in self.entities:
            if e["name"] == name:
                return e
        return None


def create_memory_server(memory_file_path: str = "./memory.jsonl") -> FastMCP:
    """Create a FastMCP server with all 9 knowledge graph tools."""
    mcp = FastMCP("jade-memory")
    graph = KnowledgeGraph(file_path=memory_file_path)
    graph.load()

    @mcp.tool()
    def create_entities(entities: list[dict[str, Any]]) -> str:
        """Create new entities in the knowledge graph."""
        created = []
        for e in entities:
            name = e.get("name", "")
            if not name:
                msg = "Entity name is required"
                raise ValueError(msg)
            existing = graph.find_entity(name)
            if existing:
                # Merge observations (preserve insertion order)
                merged = existing.get("observations", []) + e.get("observations", [])
                existing["observations"] = list(dict.fromkeys(merged))
            else:
                graph.entities.append(
                    {
                        "name": name,
                        "entityType": e.get("entityType", "Concept"),
                        "observations": e.get("observations", []),
                    }
                )
            created.append(name)
        graph.save()
        return json.dumps({"created": created})

    @mcp.tool()
    def create_relations(relations: list[dict[str, Any]]) -> str:
        """Create relations between entities."""
        created = []
        for r in relations:
            graph.relations.append(
                {
                    "from": r["from"],
                    "to": r["to"],
                    "relationType": r["relationType"],
                }
            )
            created.append(f"{r['from']} -> {r['to']}")
        graph.save()
        return json.dumps({"created": created})

    @mcp.tool()
    def add_observations(observations: list[dict[str, Any]]) -> str:
        """Add observations to existing entities."""
        added = []
        for obs in observations:
            entity_name = obs.get("entityName", "")
            contents = obs.get("contents", [])
            entity = graph.find_entity(entity_name)
            if entity:
                merged = entity.get("observations", []) + contents
                entity["observations"] = list(dict.fromkeys(merged))
                added.append({"entityName": entity_name, "addedCount": len(contents)})
        graph.save()
        return json.dumps({"added": added})

    @mcp.tool()
    def delete_entities(entityNames: list[str]) -> str:  # noqa: N803
        """Delete entities and their associated relations."""
        graph.entities = [e for e in graph.entities if e["name"] not in entityNames]
        graph.relations = [r for r in graph.relations if r["from"] not in entityNames and r["to"] not in entityNames]
        graph.save()
        return json.dumps({"deleted": entityNames})

    @mcp.tool()
    def delete_observations(deletions: list[dict[str, Any]]) -> str:
        """Delete specific observations from entities."""
        for d in deletions:
            entity_name = d.get("entityName", "")
            to_remove = set(d.get("observations", []))
            entity = graph.find_entity(entity_name)
            if entity:
                entity["observations"] = [o for o in entity.get("observations", []) if o not in to_remove]
        graph.save()
        return json.dumps({"status": "ok"})

    @mcp.tool()
    def delete_relations(relations: list[dict[str, Any]]) -> str:
        """Delete specific relations."""
        to_remove = {(r["from"], r["to"], r["relationType"]) for r in relations}
        graph.relations = [r for r in graph.relations if (r["from"], r["to"], r["relationType"]) not in to_remove]
        graph.save()
        return json.dumps({"status": "ok"})

    @mcp.tool()
    def read_graph() -> str:
        """Read the entire knowledge graph."""
        return json.dumps({"entities": graph.entities, "relations": graph.relations})

    @mcp.tool()
    def search_nodes(query: str) -> str:
        """Search for entities matching a query string."""
        query_lower = query.lower()
        matches = []
        for e in graph.entities:
            if (
                query_lower in e["name"].lower()
                or query_lower in e.get("entityType", "").lower()
                or any(query_lower in obs.lower() for obs in e.get("observations", []))
            ):
                matches.append(e)
        return json.dumps({"entities": matches, "relations": []})

    @mcp.tool()
    def open_nodes(names: list[str]) -> str:
        """Open specific nodes by name."""
        name_set = set(names)
        matches = [e for e in graph.entities if e["name"] in name_set]
        related = [r for r in graph.relations if r["from"] in name_set or r["to"] in name_set]
        return json.dumps({"entities": matches, "relations": related})

    return mcp

File: ts/mcp/memory-server.ts
/**
 * MCP Knowledge Graph Memory Server (TypeScript).
 *
 * Implements the 9-tool MCP memory protocol with JSONL persistence.
 */

import { readFileSync, writeFileSync, existsSync } from "node:fs";

interface EntityRecord {
  name: string;
  entityType: string;
  observations: string[];
}

interface RelationRecord {
  from: string;
  to: string;
  relationType: string;
}

interface KnowledgeGraph {
  entities: EntityRecord[];
  relations: RelationRecord[];
}

export interface ToolDef {
  name: string;
  description: string;
}

function loadGraph(filePath: string): KnowledgeGraph {
  const graph: KnowledgeGraph = { entities: [], relations: [] };
  if (!existsSync(filePath)) return graph;

  try {
    const content = readFileSync(filePath, "utf-8");
    for (const line of content.split("\n")) {
      if (!line.trim()) continue;
      let record: any;
      try {
        record = JSON.parse(line);
      } catch {
        continue; // Skip corrupt lines, keep loading valid ones
      }
      if (record.type === "entity") {
        const existing = graph.entities.find((e) => e.name === record.name);
        if (existing) {
          existing.observations = [
            ...new Set([...existing.observations, ...(record.observations ?? [])]),
          ];
        } else {
          graph.entities.push({
            name: record.name,
            entityType: record.entityType ?? "Concept",
            observations: record.observations ?? [],
          });
        }
      } else if (record.type === "relation") {
        graph.relations.push({
          from: record.from,
          to: record.to,
          relationType: record.relationType,
        });
      }
    }
  } catch {
    // File unreadable — start with empty graph
  }
  return graph;
}

function saveGraph(graph: KnowledgeGraph, filePath: string): void {
  const lines: string[] = [];
  for (const entity of graph.entities) {
    lines.push(JSON.stringify({ ...entity, type: "entity" }));
  }
  for (const relation of graph.relations) {
    lines.push(JSON.stringify({ ...relation, type: "relation" }));
  }
  writeFileSync(filePath, lines.join("\n") + "\n");
}

export function createMemoryServer(opts: { memoryFilePath: string }) {
  const filePath = opts.memoryFilePath;

  const tools: ToolDef[] = [
    { name: "create_entities", description: "Create new entities" },
    { name: "create_relations", description: "Create relations between entities" },
    { name: "add_observations", description: "Add observations to entities" },
    { name: "delete_entities", description: "Delete entities" },
    { name: "delete_observations", description: "Delete observations" },
    { name: "delete_relations", description: "Delete relations" },
    { name: "read_graph", description: "Read the entire graph" },
    { name: "search_nodes", description: "Search for entities" },
    { name: "open_nodes", description: "Open specific nodes" },
  ];

  return {
    async listTools(): Promise<ToolDef[]> {
      return tools;
    },

    async callTool(name: string, args: Record<string, any>): Promise<any> {
      const graph = loadGraph(filePath);

      switch (name) {
        case "create_entities": {
          const created: string[] = [];
          for (const e of args.entities ?? []) {
            if (!e.name) throw new Error("Entity name is required");
            const existing = graph.entities.find((x) => x.name === e.name);
            if (existing) {
              existing.observations = [
                ...new Set([...existing.observations, ...(e.observations ?? [])]),
              ];
            } else {
              graph.entities.push({
                name: e.name,
                entityType: e.entityType ?? "Concept",
                observations: e.observations ?? [],
              });
            }
            created.push(e.name);
          }
          saveGraph(graph, filePath);
          return { created };
        }

        case "create_relations": {
          const created: string[] = [];
          for (const r of args.relations ?? []) {
            graph.relations.push({
              from: r.from,
              to: r.to,
              relationType: r.relationType,
            });
            created.push(`${r.from} -> ${r.to}`);
          }
          saveGraph(graph, filePath);
          return { created };
        }

        case "add_observations": {
          const added: any[] = [];
          for (const obs of args.observations ?? []) {
            const entity = graph.entities.find(
              (e) => e.name === obs.entityName
            );
            if (entity) {
              entity.observations = [
                ...new Set([...entity.observations, ...(obs.contents ?? [])]),
              ];
              added.push({
                entityName: obs.entityName,
                addedCount: (obs.contents ?? []).length,
              });
            }
          }
          saveGraph(graph, filePath);
          return { added };
        }

        case "delete_entities": {
          const names = new Set(args.entityNames ?? []);
          graph.entities = graph.entities.filter((e) => !names.has(e.name));
          graph.relations = graph.relations.filter(
            (r) => !names.has(r.from) && !names.has(r.to)
          );
          saveGraph(graph, filePath);
          return { deleted: [...names] };
        }

        case "delete_observations": {
          for (const d of args.deletions ?? []) {
            const entity = graph.entities.find(
              (e) => e.name === d.entityName
            );
            if (entity) {
              const toRemove = new Set(d.observations ?? []);
              entity.observations = entity.observations.filter(
                (o) => !toRemove.has(o)
              );
            }
          }
          saveGraph(graph, filePath);
          return { status: "ok" };
        }

        case "delete_relations": {
          const toRemove = new Set(
            (args.relations ?? []).map(
              (r: any) => `${r.from}|${r.to}|${r.relationType}`
            )
          );
          graph.relations = graph.relations.filter(
            (r) => !toRemove.has(`${r.from}|${r.to}|${r.relationType}`)
          );
          saveGraph(graph, filePath);
          return { status: "ok" };
        }

        case "read_graph": {
          return {
            entities: graph.entities,
            relations: graph.relations,
          };
        }

        case "search_nodes": {
          const query = (args.query ?? "").toLowerCase();
          const matches = graph.entities.filter(
            (e) =>
              e.name.toLowerCase().includes(query) ||
              e.entityType.toLowerCase().includes(query) ||
              e.observations.some((o) => o.toLowerCase().includes(query))
          );
          return { entities: matches, relations: [] };
        }

        case "open_nodes": {
          const nameSet = new Set(args.names ?? []);
          const matches = graph.entities.filter((e) => nameSet.has(e.name));
          const related = graph.relations.filter(
            (r) => nameSet.has(r.from) || nameSet.has(r.to)
          );
          return { entities: matches, relations: related };
        }

        default:
          throw new Error(`Unknown tool: ${name}`);
      }
    },
  };
}
